Update Schema Config to include new geo fields and COG picklists

sharadsw opened the pull request:
Fixes #5215
Fixes #5289
Fixes #5292
Fixes #5293

NOTE: `COG -> cojo` has been removed and `COG -> childCojos` is renamed to `COG -> children`

> This migration updates the data model for COG and Schema Config entries for pre-geo tables and creates picklists for COGTypes.

> Data model changes:
Removed COG -> cojo
Added COG -> children

> Schema Config changes:
Added StorageTreeDef -> institution
Added COG -> children
Removed CollectionObject -> collectionObjectType (duplicate)
Removed COG -> cojo

> Creates a picklist for COGType -> type and updates an existing incorrect picklist for COG -> COGType

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add automated tests
- [ ] Add relevant issue to release milestone

### Testing instructions

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->
#### Test new geo fields
```
StorageTreeDef -> institution
CollectionObjectGroup -> children
```
- [ ] Verify above fields exist in Schema Config

#### Test COG picklists in Workbench
- Create a workbench dataset with base table CollectionObjectGroup
- Map a column to `CogType -> name` and to `CogType -> type`
- Click save
- [ ] Verify `CogType -> name` consists of all CogTypes added in the db as table records
- [ ]  Verify `CogType -> type` consists of values: Discrete, Consolidated, Drill Core

#### Test COG picklists in Data Entry
Here are some viewdefs that can be used to test the picklists in data entry

**CollectionObjectGroup**
```XML
<viewdef name="CollectionObjectGroup" class="edu.ku.brc.specify.datamodel.CollectionObjectGroup" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroup Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="cogtype" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```

**CollectionObjectGroupType**
```XML
<viewdef name="CollectionObjectGroupType" class="edu.ku.brc.specify.datamodel.CollectionObjectGroupType" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroupType Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="type" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```
- Go to Data Entry > CollectionObjectGroup
- [ ] Verify CogType is a picklist consisting of all CogTypes added in the db
- Go to Data Entry >  CollectionObjectGroupType
- [ ] Verify type is a picklist consisting of values: Discrete, Consolidated, Drill Core

- [ ] Verify a COGType and a COG can be saved 

- [ ] Verify that #5289 is fixed
- [ ] Verify that #5292 is fixed (i.e: A new CollectionObjectGroupType can be created)
- [ ] Verify that #5293 is fixed (i.e: There's a save blocker when COGType has not been selected)

sharadsw commented:
TODO: Add Sp7 tables

sharadsw commented:
TODO: Verify picklist params, check all usages of the default cogtype picklist

melton-jason commented on specifyweb/specify/migrations/0004_schema_config_update.py:
```diff
@@ -0,0 +1,177 @@
+        "field": "collectionObjectType",
+        "isrequired": False,
+    },
+    {
+        "table": "GeographyTreeDef",
+        "field": "discipline",
+        "isrequired": True,
+    },
+    {
+        "table": "GeologicTimePeriodTreeDef",
+        "field": "discipline",
+        "isrequired": True,
+    },
```
If we want to be able to create Discipline's anytime soon and keep compatibility with Specify 6, we can't mark `GeographyTreeDef -> discipline` or `GeologicTimePeriod -> discipline` as required unfortunately. 

The inverse relationships, `Discipline -> geographyTreeDef` and `Discipline -> geologicTimePeriodTreeDef` are required at the database level. 

We should actually modify the datamodel file to reflect this as well (and regenerate the frontend datamodel and testing files): 
https://github.com/specify/specify7/blob/f429d81de93184126bf42bd91ee8e6e2166cd23f/specifyweb/specify/datamodel.py#L3845
https://github.com/specify/specify7/blob/f429d81de93184126bf42bd91ee8e6e2166cd23f/specifyweb/specify/datamodel.py#L3961

https://github.com/specify/specify7/blob/f429d81de93184126bf42bd91ee8e6e2166cd23f/specifyweb/frontend/js_src/lib/components/DataModel/types.ts#L9-L17
https://github.com/specify/specify7/blob/production/specifyweb/frontend/js_src/lib/tests/ajax/static/context/datamodel.json

sharadsw replied:
What about LithostratTreeDef -> discipline and TaxonTreeDef -> discipline?

melton-jason replied:
If we want, we can leave `LithostratTreeDef -> discipline` and `TaxonTreeDef -> discipline` as required at the datamodel level (as long as it's not required at the database level). 
The `Discipline -> taxonTreeDef` and `Discipline -> lithoStratTreeDef` relationships are not required. 

```sql
MariaDB [ojsmnh]> desc discipline;
+-----------------------------+-------------+------+-----+---------+-------+
| Field                       | Type        | Null | Key | Default | Extra |
+-----------------------------+-------------+------+-----+---------+-------+
| UserGroupScopeId            | int(11)     | NO   | PRI | NULL    |       |
| TimestampCreated            | datetime    | NO   |     | NULL    |       |
| TimestampModified           | datetime    | YES  |     | NULL    |       |
| Version                     | int(11)     | YES  |     | NULL    |       |
| ModifiedByAgentID           | int(11)     | YES  | MUL | NULL    |       |
| CreatedByAgentID            | int(11)     | YES  | MUL | NULL    |       |
| disciplineId                | int(11)     | YES  |     | NULL    |       |
| Name                        | varchar(64) | YES  | MUL | NULL    |       |
| RegNumber                   | varchar(24) | YES  |     | NULL    |       |
| Type                        | varchar(64) | YES  |     | NULL    |       |
| TaxonTreeDefID              | int(11)     | YES  | MUL | NULL    |       |
| DataTypeID                  | int(11)     | NO   | MUL | NULL    |       |
| GeologicTimePeriodTreeDefID | int(11)     | NO   | MUL | NULL    |       |
| GeographyTreeDefID          | int(11)     | NO   | MUL | NULL    |       |
| DivisionID                  | int(11)     | NO   | MUL | NULL    |       |
| LithoStratTreeDefID         | int(11)     | YES  | MUL | NULL    |       |
| IsPaleoContextEmbedded      | bit(1)      | YES  |     | NULL    |       |
| PaleoContextChildTable      | varchar(50) | YES  |     | NULL    |       |
| TectonicUnitTreeDefID       | int(11)     | YES  | MUL | NULL    |       |
+-----------------------------+-------------+------+-----+---------+-------+
19 rows in set (0.022 sec)
```

Just for completeness, `Institution -> storageTreeDef` is also not required.

melton-jason commented on specifyweb/specify/migrations/0004_schema_config_update.py:
```diff
@@ -0,0 +1,177 @@
+
+    for data in FIELD_DATA:
+        Splocaleitemstr.objects.filter(
+            text=data["field"].title(),
```
Is this `.title()` intentional? Or did you mean to include a `.title()` when creating the `Splocaleitemstr`? 
The `text` here might be different from creating the Splocaleitemstr.

sharadsw replied:
I think I intended to include it when creating a Splocaleitemstr but I'll get rid of it just to keep things simpler

melton-jason commented on specifyweb/specify/migrations/0002_geo.py:
Now that the picklist created in `0002_geo` is included in an official release, directly editing the migration can still result in #5289 happening as the migration will not be reran for those who have already ran this migration. 
We will need to perform these updates in the `0004` migration,  although I don't think it necessarily hurts to keep the correct picklist definition in this previous migration as well.

sharadsw commented:
NOTE: Build/test cases will fail until #5318 is merged

emenslin reviewed (changes requested):
### Testing instructions

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->
#### Test new geo fields
```
Collection -> collectionObjectType
GeographyTreeDef -> discipline
GeologicTimePeriodTreeDef -> discipline
LithostratTreeDef -> discipline
StorageTreeDef -> institution
TaxonTreeDef -> discipline
```
- [ ] Verify above fields exist in Schema Config

All of the above except for 'storagetreedef -> institution' are duplicated in the schema results

https://github.com/user-attachments/assets/dd536e68-6aed-4036-b601-bf624b5bece1


![Screenshot 2024-10-17 101038](https://github.com/user-attachments/assets/55e1a37f-8575-4c3d-af96-8c4d6cc939c3)



#### Test COG picklists in Workbench
- Create a workbench dataset with base table CollectionObjectGroup
- Map a column to `CogType -> name` and to `CogType -> type`
- Click save
- [x] Verify `CogType -> name` consists of all CogTypes added in the db as table records
- [x]  Verify `CogType -> type` consists of values: Discrete, Consolidated, Drill Core

#### Test COG picklists in Data Entry
Here are some viewdefs that can be used to test the picklists in data entry

**CollectionObjectGroup**
```XML
<viewdef name="CollectionObjectGroup" class="edu.ku.brc.specify.datamodel.CollectionObjectGroup" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroup Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="cogtype" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```

**CollectionObjectGroupType**
```XML
<viewdef name="CollectionObjectGroupType" class="edu.ku.brc.specify.datamodel.CollectionObjectGroupType" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroupType Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="type" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```
- Go to Data Entry > CollectionObjectGroup
- [x] Verify CogType is a picklists consisting of all CogTypes added in the db
- Go to Data Entry >  CollectionObjectGroupType
- [x] Verify type is a picklist consisting of values: Discrete, Consolidated, Drill Core

- [ ] Verify that #5289 is fixed

The cogtypes picklist shows the formatter as invalid 
![Screenshot 2024-10-17 110001](https://github.com/user-attachments/assets/d6329ecb-729f-491e-bfb9-f098d78bfb02)



- [x] Verify that #5292 is fixed (i.e: A new CollectionObjectGroupType can be created)

sharadsw commented:
@emenslin Thank you! Turns out some of those fields were already added in a previous migration. I have removed the duplicate fields and changed the testing instructions. Also, the formatter being invalid is expected. The formatter will be added in #5258

emenslin reviewed (dismissed):
### Testing instructions

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->
#### Test new geo fields
```
StorageTreeDef -> institution
CollectionObjectGroup -> children
```
- [ ] Verify above fields exist in Schema Config

#### Test COG picklists in Workbench
- Create a workbench dataset with base table CollectionObjectGroup
- Map a column to `CogType -> name` and to `CogType -> type`
- Click save
- [x] Verify `CogType -> name` consists of all CogTypes added in the db as table records
- [x]  Verify `CogType -> type` consists of values: Discrete, Consolidated, Drill Core

#### Test COG picklists in Data Entry
Here are some viewdefs that can be used to test the picklists in data entry

**CollectionObjectGroup**
```XML
<viewdef name="CollectionObjectGroup" class="edu.ku.brc.specify.datamodel.CollectionObjectGroup" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroup Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="cogtype" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```

**CollectionObjectGroupType**
```XML
<viewdef name="CollectionObjectGroupType" class="edu.ku.brc.specify.datamodel.CollectionObjectGroupType" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroupType Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="type" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```
- Go to Data Entry > CollectionObjectGroup
- [x] Verify CogType is a picklist consisting of all CogTypes added in the db
- Go to Data Entry >  CollectionObjectGroupType
- [x] Verify type is a picklist consisting of values: Discrete, Consolidated, Drill Core

- [x] Verify a COGType and a COG can be saved 

- [x] Verify that #5289 is fixed
- [x] Verify that #5292 is fixed (i.e: A new CollectionObjectGroupType can be created)
- [x] Verify that #5293 is fixed (i.e: There's a save blocker when COGType has not been selected)

---

CollectionObjectGroup -> children does not show up in the schema config but it is present in the database schema. Everything else looks good though!

![Screenshot 2024-10-22 115817](https://github.com/user-attachments/assets/19d5c090-d312-4c7e-bb9d-2fa6eff621ac)

sharadsw commented:
@emenslin Looks like the migration was either skipped due to it being run already previously or incomplete on certain databases. I have manually reverted the migration on `kufish`. Should hopefully work if you create a new instance of this branch on the test panel. Can you try testing it with another db as well just in case?

pashiav reviewed (approved):
#### Test new geo fields
```
StorageTreeDef -> institution
CollectionObjectGroup -> children
```
- [X] Verify above fields exist in Schema Config

#### Test COG picklists in Workbench
- Create a workbench dataset with base table CollectionObjectGroup
- Map a column to `CogType -> name` and to `CogType -> type`
- Click save
- [X] Verify `CogType -> name` consists of all CogTypes added in the db as table records
- [X]  Verify `CogType -> type` consists of values: Discrete, Consolidated, Drill Core

#### Test COG picklists in Data Entry
Here are some viewdefs that can be used to test the picklists in data entry

**CollectionObjectGroup**
```XML
<viewdef name="CollectionObjectGroup" class="edu.ku.brc.specify.datamodel.CollectionObjectGroup" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroup Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="cogtype" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```

**CollectionObjectGroupType**
```XML
<viewdef name="CollectionObjectGroupType" class="edu.ku.brc.specify.datamodel.CollectionObjectGroupType" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The CollectionObjectGroupType Table</desc>
	<enableRules/>
	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="label" labelfor="1"/>
			<cell type="field" id="1" name="name" uitype="text"/>
			<cell type="label" labelfor="2"/>
			<cell type="field" id="2" name="type" uitype="combobox"/>
		</row>
	</rows>
</viewdef>
```
- Go to Data Entry > CollectionObjectGroup
- [X] Verify CogType is a picklist consisting of all CogTypes added in the db
- Go to Data Entry >  CollectionObjectGroupType
- [X] Verify type is a picklist consisting of values: Discrete, Consolidated, Drill Core

- [X] Verify a COGType and a COG can be saved 

- [X] Verify that #5289 is fixed
- [X] Verify that #5292 is fixed (i.e: A new CollectionObjectGroupType can be created)
- [X] Verify that #5293 is fixed (i.e: There's a save blocker when COGType has not been selected)

---
Great work! I didn't find any issues that were a part of this pr.

**Comments:**
For future @specify/ux-testing: 
1. Add the form definition for COG in App Resources, or else the `name` text field will be defaulted into a picklist because of the workbench picklist. 
- @sharadsw: "The field defaults to the workbench picklist created in this PR if the form definition does not explicitly set it as a text field." 

2. "Formatter will be invalid until we implement defaults for geo":
![image](https://github.com/user-attachments/assets/5efaa740-b955-466b-b89e-7eebc2e2e591)

3. Having multiple COGs with the same name is not reflected in the workbench picklist. Writing up an issue for this now - will update review with the link.
![image](https://github.com/user-attachments/assets/60795557-f5f9-4375-be3c-91a703a18879)
![image](https://github.com/user-attachments/assets/4ec65eaa-ce0b-4dfd-b818-366554806267)

grantfitzsimmons reviewed (approved):
> #### Test new geo fields
> ```
> StorageTreeDef -> institution
> CollectionObjectGroup -> children
> ```
> - [X] Verify above fields exist in Schema Config
> 
> #### Test COG picklists in Workbench
> - Create a workbench dataset with base table CollectionObjectGroup
> - Map a column to `CogType -> name` and to `CogType -> type`
> - Click save
> - [X] Verify `CogType -> name` consists of all CogTypes added in the db as table records
> - [X]  Verify `CogType -> type` consists of values: Discrete, Consolidated, Drill Core

This would have saved me a lot of time before the demo this morning. Nice work! 

> #### Test COG picklists in Data Entry
> Here are some viewdefs that can be used to test the picklists in data entry
> 
> **CollectionObjectGroup**
> ```XML
> <viewdef name="CollectionObjectGroup" class="edu.ku.brc.specify.datamodel.CollectionObjectGroup" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
> 	<desc>The CollectionObjectGroup Table</desc>
> 	<enableRules/>
> 	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
> 	<rowDef auto="true" cell="p" sep="2px"/>
> 	<rows>
> 		<row>
> 			<cell type="label" labelfor="1"/>
> 			<cell type="field" id="1" name="name" uitype="text"/>
> 			<cell type="label" labelfor="2"/>
> 			<cell type="field" id="2" name="cogtype" uitype="combobox"/>
> 		</row>
> 	</rows>
> </viewdef>
> ```
> 
> **CollectionObjectGroupType**
> ```XML
> <viewdef name="CollectionObjectGroupType" class="edu.ku.brc.specify.datamodel.CollectionObjectGroupType" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
> 	<desc>The CollectionObjectGroupType Table</desc>
> 	<enableRules/>
> 	<columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>
> 	<rowDef auto="true" cell="p" sep="2px"/>
> 	<rows>
> 		<row>
> 			<cell type="label" labelfor="1"/>
> 			<cell type="field" id="1" name="name" uitype="text"/>
> 			<cell type="label" labelfor="2"/>
> 			<cell type="field" id="2" name="type" uitype="combobox"/>
> 		</row>
> 	</rows>
> </viewdef>
> ```
> - Go to Data Entry > CollectionObjectGroup
> - [X] Verify CogType is a picklist consisting of all CogTypes added in the db
> - Go to Data Entry >  CollectionObjectGroupType
> - [X] Verify type is a picklist consisting of values: Discrete, Consolidated, Drill Core
> 
> - [X] Verify a COGType and a COG can be saved 
> 
> - [X] Verify that #5289 is fixed

A little different, but functional. That was the intention of that issue.

> - [X] Verify that #5292 is fixed (i.e: A new CollectionObjectGroupType can be created)
> - [X] Verify that #5293 is fixed (i.e: There's a save blocker when COGType has not been selected)

Yes, **as long as the field is on the form**, which is not unlike existing issues elsewhere (e.g. this error happens when you remove `AccessionNumber` (or any other required field) from the form and try to save)

<img width="892" alt="image" src="https://github.com/user-attachments/assets/64ed6955-f034-4569-997f-a0f4e3faf6db">

grantfitzsimmons commented:
```
specify@91f75de1bd1b:/opt/specify7$ ve/bin/python manage.py migrate specify 0006_fix_tectonic_tree_fields

Operations to perform:
  Target specific migration: 0006_fix_tectonic_tree_fields, from specify
Running migrations:
  Rendering model states... DONE
  Unapplying specify.0007_schema_config_update... OK
```
It's even reversible. 💯
